File: Airship/AirshipCore/Source/Internal/UASearchEventTemplate.m
/* Copyright Airship and Contributors */

#import "UASearchEventTemplate.h"
#import "UACustomEvent+Internal.h"

#define kUASearchEventTemplate @"search"
#define kUASearchEventTemplateCategory @"category"
#define kUASearchEventTemplateQuery @"query"
#define kUASearchEventTemplateTotalResults @"total_results"
#define kUASearchEventTemplateIdentifier @"id"
#define kUASearchEventTemplateType @"type"
#define kUASearchEventTemplateLifetimeValue @"ltv"

@interface UASearchEventTemplate()
@property (nonatomic, copy) NSString *eventName;
@end

@implementation UASearchEventTemplate

- (instancetype)initWithValue:(NSDecimalNumber *)eventValue {
    self = [super init];
    if (self) {
        self.eventName = kUASearchEventTemplate;
        self.eventValue = eventValue;
    }

    return self;
}

+ (instancetype)template {
    return [self templateWithValue:nil];
}

+ (instancetype)templateWithValue:(NSDecimalNumber *)eventValue {
    return [[self alloc] initWithValue:eventValue];
}

- (UACustomEvent *)createEvent {
    UACustomEvent *event = [UACustomEvent eventWithName:self.eventName];

    NSMutableDictionary *propertyDictionary = [NSMutableDictionary dictionary];

    if (self.eventValue) {
        [event setEventValue:self.eventValue];
        [propertyDictionary setValue:@YES forKey:kUASearchEventTemplateLifetimeValue];
    } else {
        [propertyDictionary setValue:@NO forKey:kUASearchEventTemplateLifetimeValue];
    }
    
    if (self.identifier) {
        [propertyDictionary setValue:self.identifier forKey:kUASearchEventTemplateIdentifier];
    }

    if (self.category) {
        [propertyDictionary setValue:self.category forKey:kUASearchEventTemplateCategory];
    }

    if (self.query) {
        [propertyDictionary setValue:self.query forKey:kUASearchEventTemplateQuery];
    }
    
    if (self.type) {
        [propertyDictionary setValue:self.type forKey:kUASearchEventTemplateType];
    }

    if (self.totalResults) {
        [propertyDictionary setValue:@(self.totalResults) forKey:kUASearchEventTemplateTotalResults];
    }
    
    event.templateType = kUASearchEventTemplate;
    
    event.properties = propertyDictionary;

    return event;
}

@end

File: Airship/AirshipCore/Source/UASearchEventTemplate.swift
/* Copyright Airship and Contributors */

/**
 * A UASearchEventTemplate represents a custom search event template for the
 * application.
 */
@objc
public class UASearchEventTemplate : NSObject {
    /**
     * The event's value. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     */
    @objc
    public var eventValue: NSNumber?

    /**
     * The event's type.
     */
    @objc
    public var type: String?

    /**
     * The event's identifier.
     */
    @objc
    public var identifier: String?


    /**
     * The event's category.
     */
    @objc
    public var category: String?

    /**
     * The event's query.
     */
    @objc
    public var query: String?

    /**
     * The event's total results.
     */
    @objc
    public var totalResults: Int = 0


    /**
     * Default constructor.
     *
     * @param value The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     */
    public init(value: NSNumber? = nil) {
        super.init()
        self.eventValue = value
    }

    /**
     * Factory method for creating a search event template with a value.
     *
     * @param value The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * @return UASearchEventTemplate instance.
     */
    @objc(templateWithValue:)
    public class func template(value: NSNumber) -> UASearchEventTemplate {
        return UASearchEventTemplate(value: value)
    }

    /**
     * Factory method for creating a search event template.
     * @return UASearchEventTemplate instance.
     */
    @objc(template)
    public class func template() -> UASearchEventTemplate {
        return UASearchEventTemplate()
    }

    /**
     * Creates the custom search event.
     * @return Created UACustomEvent instance.
     */
    @objc
    public func createEvent() -> UACustomEvent {
        var propertyDictionary: [String : Any] = [:]
        propertyDictionary["ltv"] = self.eventValue != nil
        propertyDictionary["id"] = identifier
        propertyDictionary["category"] = category
        propertyDictionary["query"] = query
        propertyDictionary["type"] = type
        propertyDictionary["total_results"] = self.totalResults > 0 ? self.totalResults : nil

        let event = UACustomEvent(name: "search")
        event.eventValue = self.eventValue
        event.templateType = "search"
        event.properties = propertyDictionary
        return event
    }
}
